### PR TITLE
Enable support for objects, loosen validation for literals

### LINKED ISSUE #25: Inline object notation
The readme says this library doesn’t have a way to represent JSON object literals for the `literal` expression, due to a `jsep` limitation. Perhaps there could be an inline notation such as `object(key1, value1, key2, value2, …)`.

There are a few error conditions to consider:

* Duplicate keys
* An odd number of arguments to the function
* Keys that aren’t strings

Besides `literal`, an object literal syntax is needed because mapbox/mapbox-gl-js#6270 introduced a `Collator` object type that can be passed into various comparison expressions.

/cc @dasulit @davidtheclark @samanpwbb @anandthakker @ChrisLoer

### LINKED ISSUE #33: Literal expression validation is too strict
We have some validation for literal expressions that only allow arrays as the second argument, ie

`[ "literal", [ 1, 2, 3 ] ]`

Currently, a primitive value will pass validation, ie

`[ "literal", "#666666" ]`

Upstream: https://github.com/mapbox/mapbox-gl-js/issues/9308

Depending on the resolution of that ticket, we may need to relax our validation here.

### LINKED ISSUE #34: Add support for index-of, slice, distance, and within expressions
There are new expressions introduced since GL 1.9 that need support in expression jamsession. 

https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#index-of
https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#slice
https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#distance
https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#within

### PR DESCRIPTION
**Pull request by dereklieu**
- Enables object support, closes #25 
- Loosens validation around literals, closes #33 
- Bumps gl style spec, closes #34 
- Bumps up a few build dependency versions

Apologies in advance for bundling a couple things into this PR!

### CODE REVIEW

File: src/expression-to-formula.js [hunk 1]
```diff
@@ -26,19 +26,21 @@ export default function expressionToFormula(expression) {
...
 
   const args = expression.slice(1).map(arg => {
     if (typeof arg === 'string') {
       return `"${arg}"`;
     }
-    if (!Array.isArray(arg)) {
+
+    const isArray = Array.isArray(arg);
+    if (typeof arg === 'object' && !!arg && !isArray) {
+      return JSON.stringify(arg);
```
Comment (dereklieu) [inline src/expression-to-formula.js:41]: 🤔  maybe we should do some pretty-printing here? Maybe add spaces?

Comment (samanpwbb) [reply src/expression-to-formula.js:41]: spaces!

File: test/formula-to-expression.test.js [hunk 2]
```diff
@@ -246,6 +265,39 @@ describe('formulas', () => {
...
+    );
+    expect(actual).toEqual([
+      'number-format',
+      1.005,
+      {
+        'max-fraction-digits': 2,
+        'min-fraction-digits': 2
+      }
+    ]);
+  });
```
Comment (samanpwbb) [inline test/formula-to-expression.test.js:280]: 🎉

**Review by samanpwbb — APPROVED** [review]
Looks great! I think we can remove the blocklist here: https://github.com/mapbox/expression-jamsession/blob/master/scripts/build-operator-whitelist.js#L8-L14

**Pull request merged**